Sources/APIServer/Volumes/VolumesService.swift

```diff
@@ -0,0 +1,220 @@
+//===----------------------------------------------------------------------===//
+// Copyright © 2025 Apple Inc. and the container project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+
+import ContainerClient
+import ContainerPersistence
+import Containerization
+import ContainerizationEXT4
+import ContainerizationError
+import ContainerizationExtras
+import ContainerizationOS
+import Foundation
+import Logging
+import Synchronization
+import SystemPackage
+
+actor VolumesService {
+    private let resourceRoot: URL
+    private let store: ContainerPersistence.FilesystemEntityStore<Volume>
+    private let log: Logger
+    private let lock = AsyncLock()
+    private weak var containersService: ContainersService?
+
+    // Storage constants
+    private static let entityFile = "entity.json"
+    private static let blockFile = "volume.img"
+
+    public init(resourceRoot: URL, log: Logger) throws {
+        try FileManager.default.createDirectory(at: resourceRoot, withIntermediateDirectories: true)
+        self.resourceRoot = resourceRoot
+        self.store = try FilesystemEntityStore<Volume>(path: resourceRoot, type: "volumes", log: log)
+        self.log = log
+    }
+
+    public func setContainersService(_ containersService: ContainersService) {
+        self.containersService = containersService
+    }
+
+    private func parseSize(_ sizeString: String) throws -> UInt64 {
+        let measurement = try Measurement.parse(parsing: sizeString)
+        let bytes = measurement.converted(to: .bytes).value
+
+        // Validate reasonable size limits
+        let minSize: UInt64 = 1.mib()  // 1mib minimum
+        let maxSize: UInt64 = 1.tib()  // 1tib maximum
+
+        let sizeInBytes = UInt64(bytes)
+
+        guard sizeInBytes >= minSize else {
+            throw VolumeError.storageError("Volume size too small: minimum 1MB")
```

Comment:
Going to be pedantic, but:
```suggestion
            throw VolumeError.storageError("Volume size too small: minimum 1MiB")
```